config/confighttp/confighttp.go

```diff
@@ -342,15 +342,18 @@ func (hss *HTTPServerSettings) ToServer(host component.Host, settings component.
 		handler = authInterceptor(handler, server)
 	}
 
-	// TODO: emit a warning when non-empty CorsHeaders and empty CorsOrigins.
-	if hss.CORS != nil && len(hss.CORS.AllowedOrigins) > 0 {
-		co := cors.Options{
-			AllowedOrigins:   hss.CORS.AllowedOrigins,
-			AllowCredentials: true,
-			AllowedHeaders:   hss.CORS.AllowedHeaders,
-			MaxAge:           hss.CORS.MaxAge,
+	if hss.CORS != nil {
+		if len(hss.CORS.AllowedOrigins) > 0 {
+			co := cors.Options{
+				AllowedOrigins:   hss.CORS.AllowedOrigins,
+				AllowCredentials: true,
+				AllowedHeaders:   hss.CORS.AllowedHeaders,
+				MaxAge:           hss.CORS.MaxAge,
+			}
+			handler = cors.New(co).Handler(handler)
+		} else if len(hss.CORS.AllowedHeaders) > 0 {
+			settings.Logger.Warn("The CORS configuration specifies allowed headers but no allowed origins, and is therefore ignored.")
```

Comment:
Super hard to review, can you try to refactor this code?

Reply:
ok sure, I made a much smaller diff, please feel free to review